shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java

```diff
@@ -23,6 +42,50 @@ public ClassLoader getSplitClassLoader(String splitName) throws NameNotFoundExce
     return this.getClass().getClassLoader();
   }
 
+  private void tryInitAppComponentFactory(LoadedApk realLoadedApk) {
+    if (RuntimeEnvironment.getApiLevel() >= VERSION_CODES.P) {
+      ApplicationInfo applicationInfo = realLoadedApk.getApplicationInfo();
+      if (applicationInfo == null || applicationInfo.appComponentFactory == null) {
+        return;
+      }
+      _LoadedApk_ loadedApkReflector = reflector(_LoadedApk_.class, realLoadedApk);
+      if (!loadedApkReflector.getIncludeCode()) {
+        return;
+      }
+      String fullQualifiedClassName =
+          replaceLastDotWith$IfInnerStaticClass(
+              calculateFullQualifiedClassName(
+                  applicationInfo.appComponentFactory, applicationInfo.packageName));
+      android.app.AppComponentFactory factory =
+          (android.app.AppComponentFactory) newInstanceOf(fullQualifiedClassName);
+      if (factory == null) {
+        factory = new AppComponentFactory();
+      }
+      loadedApkReflector.setAppFactory(factory);
+    }
+  }
+
+  private static String replaceLastDotWith$IfInnerStaticClass(String receiverClassName) {
```

Comment:
If you remove this method do the tests still pass?

Reply:
I removed this method from AndroidTestEnvironment and everything still works:

https://github.com/robolectric/robolectric/commit/40c16eaac4bbd63069ce8a41df681afc6b8561c3

Reply:
I think it can work without this method. The reason that I copy it here is to keep the consistent with `AndroidTestEnvironment`. The critical part is to resolve the package name for class with the sytle `.TestClass`. I also will check whether we can remove it too because we use the single `ClassLoader` of the Robolectric.

Reply:
Yeah I confirmed that it works without this method, so it can be removed.

Reply:
Updated, and let's check the CI result.